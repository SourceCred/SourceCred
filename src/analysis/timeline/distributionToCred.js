// @flow

/**
 * Takes timeline distributions and uses them to create normalized cred.
 */

import {sum} from "d3-array";
import {type Interval} from "./interval";
import {type TimelineDistributions} from "./timelinePagerank";
import {NodeAddress, type NodeAddressT} from "../../core/graph";

/**
 * Represents the full timeline cred for a graph.
 */
export type FullTimelineCred = $ReadOnlyArray<{|
  // The interval for this slice.
  +interval: Interval,
  // The cred for each node.
  // (Uses the graph's canonical node ordering.)
  +cred: Float64Array,
|}>;

/**
 * Convert a TimelineDistribution into TimelineCred.
 *
 * The difference between the distribution and cred is that cred has been
 * re-normalized to present human-agreeable scores, rather than a probability
 * distribution.
 *
 * This implementation normalizes the scores so that in each interval, the
 * total score of every node matching scoringNodePrefix is equal to the
 * interval's weight.
 *
 * Edge cases:
 *     - If in an interval the sum of the scoring nodes' distribution is 0,
 *       return a total cred score of 0 for all nodes in the interval.
 *     - If none of the nodes match a scoring node prefix, return
 *       a total cred score of 0 for all nodes in all intervals.
 *
 */
export function distributionToCred(
  ds: TimelineDistributions,
  nodeOrder: $ReadOnlyArray<NodeAddressT>,
  scoringNodePrefixes: $ReadOnlyArray<NodeAddressT>
): FullTimelineCred {
  if (ds.length === 0) {
    return [];
  }
  const intervals = ds.map((x) => x.interval);
  const scoringNodeIndices = [];
  const cred = new Array(nodeOrder.length);
  for (let i = 0; i < nodeOrder.length; i++) {
    const addr = nodeOrder[i];
    if (scoringNodePrefixes.some((x) => NodeAddress.hasPrefix(addr, x))) {
      scoringNodeIndices.push(i);
    }
    cred[i] = new Array(intervals.length);
  }

  return ds.map(({interval, distribution, intervalWeight}) => {
    const intervalTotalScore = sum(
      scoringNodeIndices.map((x) => distribution[x])
    );

    const intervalNormalizer =
      intervalTotalScore == 0 ? 0 : intervalWeight / intervalTotalScore;
    const cred = distribution.map((x) => x * intervalNormalizer);
    return {interval, cred};
  });
}
