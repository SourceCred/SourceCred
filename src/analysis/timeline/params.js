// @flow

import * as C from "../../util/combo";
import * as N from "../../util/numerics";

/**
 * Parameters for computing TimelineCred
 *
 * The parameters are intended to be user-configurable.
 */
export type TimelineCredParameters = {|
  // Determines how quickly cred returns to the PageRank seed vector. If alpha
  // is high, then cred will tend to "stick" to nodes that are seeded, e.g.
  // issues and pull requests. Alpha should be between 0 and 1.
  +alpha: N.Proportion,
  // Determines how quickly cred decays. The decay is 1, then cred never
  // decays, and old nodes and edges will retain full weight forever. (This
  // would result in cred that is highly biased towards old contributions, as
  // they would continue earning cred in every timeslice, forever.) If the
  // decay is 0, then weights go to zero the first week after their node/edge
  // was created. Should be between 0 and 1.
  +intervalDecay: N.Proportion,
|};

export const DEFAULT_ALPHA = N.proportion(0.2);
export const DEFAULT_INTERVAL_DECAY = N.proportion(0.5);

export type TimelineCredParametersJSON = {|
  +alpha: number,
  +intervalDecay: number,
|};

export function paramsToJSON(
  p: TimelineCredParameters
): TimelineCredParametersJSON {
  return {
    alpha: p.alpha,
    intervalDecay: p.intervalDecay,
  };
}

export function paramsFromJSON(
  p: TimelineCredParametersJSON
): TimelineCredParameters {
  return {
    alpha: N.proportion(p.alpha),
    intervalDecay: N.proportion(p.intervalDecay),
  };
}

const partialParser: C.Parser<$Shape<TimelineCredParameters>> = C.shape({
  alpha: N.proportionParser,
  intervalDecay: N.proportionParser,
});

export const parser: C.Parser<TimelineCredParameters> = C.fmap(
  partialParser,
  partialParams
);

/**
 * Exports the default TimelineCredParameters.
 *
 * End consumers of SourceCred will not need to depend on this; it's
 * provided for implementation of SourceCred's APIs.
 */
export function defaultParams(): TimelineCredParameters {
  return {
    alpha: DEFAULT_ALPHA,
    intervalDecay: DEFAULT_INTERVAL_DECAY,
  };
}

/**
 * Fill in default values for timeline cred parameters.
 */
export function partialParams(
  partial: $Shape<TimelineCredParameters> | null
): TimelineCredParameters {
  return {...defaultParams(), ...partial};
}
