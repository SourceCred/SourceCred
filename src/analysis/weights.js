// @flow

import * as MapUtil from "../util/map";
import {type NodeAddressT, type EdgeAddressT} from "../core/graph";
import type {EdgeType, NodeType} from "./types";
import type {PluginDeclaration} from "./pluginDeclaration";

export type WeightedNodeType = {|+type: NodeType, +weight: number|};

export type WeightedEdgeType = {|
  +type: EdgeType,
  +forwardWeight: number,
  +backwardWeight: number,
|};

export type WeightedTypes = {|
  // Map from the weighted type's prefix to the type
  +nodes: Map<NodeAddressT, WeightedNodeType>,
  +edges: Map<EdgeAddressT, WeightedEdgeType>,
|};

export function defaultWeightedNodeType(type: NodeType): WeightedNodeType {
  return {
    type,
    weight: type.defaultWeight,
  };
}

export function defaultWeightedEdgeType(type: EdgeType): WeightedEdgeType {
  return {
    type,
    forwardWeight: type.defaultForwardWeight,
    backwardWeight: type.defaultBackwardWeight,
  };
}

export function defaultWeightsForDeclaration(
  declaration: PluginDeclaration
): WeightedTypes {
  return {
    nodes: new Map(
      declaration.nodeTypes.map((x) => [x.prefix, defaultWeightedNodeType(x)])
    ),
    edges: new Map(
      declaration.edgeTypes.map((x) => [x.prefix, defaultWeightedEdgeType(x)])
    ),
  };
}

export function combineWeights(
  ws: $ReadOnlyArray<WeightedTypes>
): WeightedTypes {
  return {
    nodes: MapUtil.merge(ws.map((x) => x.nodes)),
    edges: MapUtil.merge(ws.map((x) => x.edges)),
  };
}
