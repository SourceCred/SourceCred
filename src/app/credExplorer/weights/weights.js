// @flow

import * as MapUtil from "../../../util/map";
import {type NodeAddressT, type EdgeAddressT} from "../../../core/graph";
import type {NodeType, EdgeType} from "../../adapters/pluginAdapter";
import type {StaticPluginAdapter} from "../../adapters/pluginAdapter";
import type {StaticAdapterSet} from "../../adapters/adapterSet";

export type WeightedNodeType = {|+type: NodeType, +weight: number|};

export type WeightedEdgeType = {|
  +type: EdgeType,
  +forwardWeight: number,
  +backwardWeight: number,
|};

export type WeightedTypes = {|
  // Map from the weighted type's prefix to the type
  +nodes: Map<NodeAddressT, WeightedNodeType>,
  +edges: Map<EdgeAddressT, WeightedEdgeType>,
|};

export function defaultWeightedNodeType(type: NodeType): WeightedNodeType {
  return {
    type,
    weight: type.defaultWeight,
  };
}

export function defaultWeightedEdgeType(type: EdgeType): WeightedEdgeType {
  return {
    type,
    forwardWeight: type.defaultForwardWeight,
    backwardWeight: type.defaultBackwardWeight,
  };
}

export function defaultWeightsForAdapter(
  adapter: StaticPluginAdapter
): WeightedTypes {
  return {
    nodes: new Map(
      adapter.nodeTypes().map((x) => [x.prefix, defaultWeightedNodeType(x)])
    ),
    edges: new Map(
      adapter.edgeTypes().map((x) => [x.prefix, defaultWeightedEdgeType(x)])
    ),
  };
}

export function combineWeights(
  ws: $ReadOnlyArray<WeightedTypes>
): WeightedTypes {
  return {
    nodes: MapUtil.merge(ws.map((x) => x.nodes)),
    edges: MapUtil.merge(ws.map((x) => x.edges)),
  };
}

export function defaultWeightsForAdapterSet(
  adapters: StaticAdapterSet
): WeightedTypes {
  return combineWeights(adapters.adapters().map(defaultWeightsForAdapter));
}
