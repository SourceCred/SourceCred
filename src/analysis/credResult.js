// @flow

import {
  type WeightedGraph,
  type WeightedGraphJSON,
  toJSON as wgToJSON,
  fromJSON as wgFromJSON,
} from "../core/weightedGraph";
import {NodeAddress} from "../core/graph";
import {type Compatible, toCompat, fromCompat} from "../util/compat";
import {
  type TimelineCredParameters,
  type TimelineCredParametersJSON,
  paramsToJSON,
  paramsFromJSON,
} from "./timeline/params";
import {
  type PluginDeclarations,
  type PluginDeclarationsJSON,
  toJSON as pluginsToJSON,
  fromJSON as pluginsFromJSON,
} from "./pluginDeclaration";
import {timelinePagerank} from "../core/algorithm/timelinePagerank";
import {
  type CredData,
  computeCredData,
  compressByThreshold as _compressByThreshold,
  compressDownToMatchingIndices,
} from "./credData";
import {distributionToCred} from "../core/algorithm/distributionToCred";

/**
 * Comprehensive cred output data, including the graph, the scores, the params, and the plugins.
 */
export type CredResult = {|
  // The Graph on which this cred data was computed.
  +weightedGraph: WeightedGraph,
  // Cred level information, always stored in graph address order.
  +credData: CredData,
  // Parameters used to compute this cred data
  +params: TimelineCredParameters,
  // Plugin declarations used to compute this cred data
  +plugins: PluginDeclarations,
|};

export async function compute(
  wg: WeightedGraph,
  params: TimelineCredParameters,
  plugins: PluginDeclarations
): Promise<CredResult> {
  const {graph} = wg;
  const nodeOrder = Array.from(graph.nodes()).map((x) => x.address);
  const userTypes = [].concat(...plugins.map((x) => x.userTypes));
  const scorePrefixes = userTypes.map((x) => x.prefix);
  const distribution = await timelinePagerank(
    wg,
    params.intervalDecay,
    params.alpha
  );
  const credScores = distributionToCred(distribution, nodeOrder, scorePrefixes);
  const credData = computeCredData(credScores);
  return {weightedGraph: wg, credData, params, plugins};
}

// Lossily compress a CredResult, by throwing away time-specific cred
// data for any flows that summed to less than `threshold` cred.
// We may want to implement more sophisticated and context-aware strategies
// in the future.
export function compressByThreshold(
  x: CredResult,
  threshold: number
): CredResult {
  const {params, plugins, weightedGraph, credData} = x;
  return {
    params,
    plugins,
    weightedGraph,
    credData: _compressByThreshold(credData, threshold),
  };
}

export function stripOverTimeDataForNonUsers(x: CredResult): CredResult {
  const {params, plugins, weightedGraph, credData} = x;
  const userPrefixes = []
    .concat(...plugins.map((x) => x.userTypes))
    .map((x) => x.prefix);
  const nodeOrder = Array.from(weightedGraph.graph.nodes()).map(
    (x) => x.address
  );
  const inclusionIndices = new Set();
  nodeOrder.forEach((a, i) => {
    if (userPrefixes.some((p) => NodeAddress.hasPrefix(a, p))) {
      inclusionIndices.add(i);
    }
  });
  const newCredData = compressDownToMatchingIndices(credData, inclusionIndices);
  return {params, plugins, weightedGraph, credData: newCredData};
}

const COMPAT_INFO = {type: "sourcecred/credResult", version: "0.1.0"};

export type CredResultJSON = Compatible<{|
  +weightedGraph: WeightedGraphJSON,
  +credData: CredData,
  +params: TimelineCredParametersJSON,
  +plugins: PluginDeclarationsJSON,
|}>;

export function toJSON(x: CredResult): CredResultJSON {
  const {weightedGraph, credData, params, plugins} = x;
  return toCompat(COMPAT_INFO, {
    weightedGraph: wgToJSON(weightedGraph),
    credData,
    params: paramsToJSON(params),
    plugins: pluginsToJSON(plugins),
  });
}

export function fromJSON(j: CredResultJSON): CredResult {
  const x = fromCompat(COMPAT_INFO, j);
  const {weightedGraph, credData, params, plugins} = x;
  return {
    weightedGraph: wgFromJSON(weightedGraph),
    credData,
    params: paramsFromJSON(params),
    plugins: pluginsFromJSON(plugins),
  };
}
