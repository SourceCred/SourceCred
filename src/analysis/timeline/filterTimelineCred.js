// @flow

import {type Interval} from "./interval";
import {NodeAddress, type NodeAddressT} from "../../core/graph";
import {type FullTimelineCred} from "./distributionToCred";

export type FilteredTimelineCred = {|
  +intervals: $ReadOnlyArray<Interval>,
  +addressToCred: Map<NodeAddressT, $ReadOnlyArray<number>>,
|};

/**
 * Compress FullTimelineCred by discarding most nodes' cred.
 *
 * FullTimelineCred contains the cred at every interval for every node in the
 * graph. This could be tens of thousands of nodes and hundreds of intervals;
 * it's ungainly to store. To avoid this issue, we compress the cred down by
 * removing cred for most nodes. (We care a lot about users' cred; not so much
 * about the cred for every individual comment ever.)
 *
 * Right now, we do this by filtering out every node that doesn't match an
 * inclusion address prefix. In the future, we may have more sophisticated
 * logic, like keeping the top k nodes for each type.
 */
export function filterTimelineCred(
  fullCred: FullTimelineCred,
  nodeOrder: $ReadOnlyArray<NodeAddressT>,
  inclusionPrefixes: $ReadOnlyArray<NodeAddressT>
): FilteredTimelineCred {
  const intervals = fullCred.map((x) => x.interval);
  const addressToCred = new Map();
  function hasMatch(x: NodeAddressT): boolean {
    for (const prefix of inclusionPrefixes) {
      if (NodeAddress.hasPrefix(x, prefix)) {
        return true;
      }
    }
    return false;
  }
  for (let i = 0; i < nodeOrder.length; i++) {
    const addr = nodeOrder[i];
    if (hasMatch(addr)) {
      const addrCred = fullCred.map(({cred}) => cred[i]);
      addressToCred.set(addr, addrCred);
    }
  }
  return {intervals, addressToCred};
}
