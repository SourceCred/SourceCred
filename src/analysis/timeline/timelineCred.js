// @flow

import {sum} from "d3-array";
import sortBy from "lodash.sortby";
import * as NullUtil from "../../util/null";
import * as MapUtil from "../../util/map";
import {toCompat, fromCompat, type Compatible} from "../../util/compat";
import {type Interval} from "./interval";
import {timelinePagerank} from "./timelinePagerank";
import {distributionToCred} from "./distributionToCred";
import {
  Graph,
  type GraphJSON,
  type NodeAddressT,
  NodeAddress,
  type Node,
} from "../../core/graph";
import {
  type Weights,
  type WeightsJSON,
  toJSON as weightsToJSON,
  fromJSON as weightsFromJSON,
} from "../weights";
import {type NodeAndEdgeTypes} from "../types";
import {
  filterTimelineCred,
  type FilteredTimelineCred,
} from "./filterTimelineCred";

export type {Interval} from "./interval";

/**
 * A Graph Node wrapped with cred information.
 */
export type CredNode = {|
  // The Graph Node in question
  +node: Node,
  // The total aggregated cred. (Summed across every interval).
  +total: number,
  // The timeline sequence of cred (one score per interval).
  +cred: $ReadOnlyArray<number>,
|};

/**
 * Parameters for computing TimelineCred
 *
 * The parameters are intended to be user-configurable.
 */
export type TimelineCredParameters = {|
  // Determines how quickly cred returns to the PageRank seed vector. If alpha
  // is high, then cred will tend to "stick" to nodes that are seeded, e.g.
  // issues and pull requests. Alpha should be between 0 and 1.
  +alpha: number,
  // Determines how quickly cred decays. The decay is 1, then cred never
  // decays, and old nodes and edges will retain full weight forever. (This
  // would result in cred that is highly biased towards old contributions, as
  // they would continue earning cred in every timeslice, forever.) If the
  // decay is 0, then weights go to zero the first week after their node/edge
  // was created. Should be between 0 and 1.
  +intervalDecay: number,
  // The weights. This determines how much cred is assigned based on different
  // node types, how cred flows across various edge types, and can specify
  // manual weights directly on individual nodes. See the docs in
  // `analysis/weights` for details.
  +weights: Weights,
|};

/**
 * Configuration for computing TimelineCred
 *
 * Unlike the parameters, the config is expected to be static.
 * It's code-level config that isolates the TimelineCred algorithms from
 * specific plugin-level details about which nodes addresses are used for scoring,
 * etc.
 *
 * A default config is available in `src/plugins/defaultCredConfig`
 */
export type TimelineCredConfig = {|
  // Cred is normalized so that for a given interval, the total score of all
  // nodes matching this prefix will be equal to the total weight of nodes in
  // the interval.
  +scoreNodePrefix: NodeAddressT,
  // To save on space, we keep cred only for nodes matching one of these
  // NodeAddresses.
  +filterNodePrefixes: $ReadOnlyArray<NodeAddressT>,
  // The types are used to assign base cred to nodes based on their type. Node
  // that the weight for each type may be overriden in the params.
  +types: NodeAndEdgeTypes,
|};

/**
 * Represents the timeline cred of a graph. This class wraps all the data
 * needed to analyze and interpet cred (ie. it has the Graph and the cred
 * scores), and provides convenient view methods for accessing the cred.
 *
 * The TimelineCred also has the params and config. The intention is that this
 * is a "one stop shop" for serializing SourceCred results.
 */
export class TimelineCred {
  _graph: Graph;
  _intervals: $ReadOnlyArray<Interval>;
  _addressToCred: Map<NodeAddressT, $ReadOnlyArray<number>>;
  _params: TimelineCredParameters;
  _config: TimelineCredConfig;

  constructor(
    graph: Graph,
    intervals: $ReadOnlyArray<Interval>,
    addressToCred: Map<NodeAddressT, $ReadOnlyArray<number>>,
    params: TimelineCredParameters,
    config: TimelineCredConfig
  ) {
    this._graph = graph;
    this._intervals = intervals;
    this._addressToCred = addressToCred;
    this._params = params;
    this._config = config;
  }

  graph(): Graph {
    return this._graph;
  }

  params(): TimelineCredParameters {
    return this._params;
  }

  config(): TimelineCredConfig {
    return this._config;
  }

  /**
   * Creates a new TimelineCred based on the new Parameters.
   * Holds the graph and config constant.
   *
   * This returns a new TimelineCred; it does not modify the existing one.
   */
  async reanalyze(newParams: TimelineCredParameters): Promise<TimelineCred> {
    return await TimelineCred.compute(this._graph, newParams, this._config);
  }

  /**
   * Return all the intervals in the timeline.
   */
  intervals(): $ReadOnlyArray<Interval> {
    return this._intervals;
  }

  /**
   * Get the CredNode for a given NodeAddress.
   *
   * Returns undefined if the node is not in the filtered results.
   *
   * Note that it's possible that the node is present in the Graph, but not the
   * filtered results; if so, it will return undefined.
   */
  credNode(a: NodeAddressT): ?CredNode {
    const cred = this._addressToCred.get(a);
    if (cred === undefined) {
      return undefined;
    }
    const total = sum(cred);
    const node = NullUtil.get(this._graph.node(a));
    return {cred, total, node};
  }

  /**
   * Return all the nodes matching the prefix, along with their cred,
   * sorted by total cred (descending).
   */
  credSortedNodes(prefix: NodeAddressT): $ReadOnlyArray<CredNode> {
    const match = (a) => NodeAddress.hasPrefix(a, prefix);
    const addresses = Array.from(this._addressToCred.keys()).filter(match);
    const credNodes = addresses.map((a) => this.credNode(a));
    return sortBy(credNodes, (x: CredNode) => -x.total);
  }

  toJSON(): TimelineCredJSON {
    const rawJSON = {
      graphJSON: this._graph.toJSON(),
      intervalsJSON: this._intervals,
      credJSON: MapUtil.toObject(this._addressToCred),
      paramsJSON: paramsToJSON(this._params),
      configJSON: this._config,
    };
    return toCompat(COMPAT_INFO, rawJSON);
  }

  static fromJSON(j: TimelineCredJSON): TimelineCred {
    const json = fromCompat(COMPAT_INFO, j);
    const {graphJSON, intervalsJSON, credJSON, paramsJSON, configJSON} = json;
    const cred = MapUtil.fromObject(credJSON);
    const graph = Graph.fromJSON(graphJSON);
    const params = paramsFromJSON(paramsJSON);
    return new TimelineCred(graph, intervalsJSON, cred, params, configJSON);
  }

  static async compute(
    graph: Graph,
    params: TimelineCredParameters,
    config: TimelineCredConfig
  ): Promise<TimelineCred> {
    const ftc = await _computeTimelineCred(graph, params, config);
    return new TimelineCred(
      graph,
      ftc.intervals,
      ftc.addressToCred,
      params,
      config
    );
  }
}

async function _computeTimelineCred(
  graph: Graph,
  params: TimelineCredParameters,
  config: TimelineCredConfig
): Promise<FilteredTimelineCred> {
  const nodeOrder = Array.from(graph.nodes()).map((x) => x.address);
  const distribution = await timelinePagerank(
    graph,
    config.types,
    params.weights,
    params.intervalDecay,
    params.alpha
  );
  const cred = distributionToCred(
    distribution,
    nodeOrder,
    config.scoreNodePrefix
  );
  const filtered = filterTimelineCred(
    cred,
    nodeOrder,
    config.filterNodePrefixes
  );
  return filtered;
}

const COMPAT_INFO = {type: "sourcecred/timelineCred", version: "0.2.0"};

export opaque type TimelineCredJSON = Compatible<{|
  +graphJSON: GraphJSON,
  +paramsJSON: ParamsJSON,
  +configJSON: TimelineCredConfig,
  +credJSON: {[string]: $ReadOnlyArray<number>},
  +intervalsJSON: $ReadOnlyArray<Interval>,
|}>;

type ParamsJSON = {|
  +alpha: number,
  +intervalDecay: number,
  +weights: WeightsJSON,
|};

function paramsToJSON(p: TimelineCredParameters): ParamsJSON {
  return {
    alpha: p.alpha,
    intervalDecay: p.intervalDecay,
    weights: weightsToJSON(p.weights),
  };
}

function paramsFromJSON(p: ParamsJSON): TimelineCredParameters {
  return {
    alpha: p.alpha,
    intervalDecay: p.intervalDecay,
    weights: weightsFromJSON(p.weights),
  };
}
