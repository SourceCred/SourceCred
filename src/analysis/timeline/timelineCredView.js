// @flow

import {sum} from "d3-array";
import sortBy from "lodash.sortby";
import stringify from "json-stable-stringify";
import * as NullUtil from "../../util/null";
import {TimelineCred} from "./timelineCred";
import {type Interval, weekIntervals} from "./interval";
import {type Node, type NodeAddressT, NodeAddress} from "../../core/graph";

/**
 * A Graph Node wrapped with cred information.
 */
export type CredNode = {|
  // The Graph Node in question
  +node: Node,
  // The total aggregated cred. (Summed across every interval).
  +total: number,
  // The timeline sequence of cred (one score per interval).
  +cred: $ReadOnlyArray<number>,
|};

/**
 * TimelineCredView allows a retrieving information about nodes' cred
 * for a specific interval of time. The TimelineCredView wraps a TimelineCred
 * instance, and provides methods like `credNode` and `credSortedNodes` for
 * retrieving information on the available nodes and how much cred they earned.
 * Every TimelineCredView is scoped to a certain view interval, which defines
 * the min and max time of cred being viewed. When `CredNode` results are returned,
 * the `total` field will correspond to the view interval of the TimelineCredView.
 * By default, the view interval is chosen to include every underlying interval
 * in the TimelineCred`.
 */
export class TimelineCredView {
  _inputInterval: Interval;
  _viewIntervals: Interval[];
  _timelineCred: TimelineCred;

  // We are displaying a range of the intervals in the base TimelineCred
  // We track the start index and end index of the range, so we can efficiently
  // slice the cred
  _startIntervalIndex: number;
  _endIntervalIndex: number;

  constructor(timelineCred: TimelineCred, inputInterval: ?Interval) {
    this._timelineCred = timelineCred;
    this._inputInterval = NullUtil.orElse(
      inputInterval,
      this._getDefaultInterval(timelineCred)
    );

    this._viewIntervals = weekIntervals(
      this._inputInterval.startTimeMs,
      this._inputInterval.endTimeMs
    );

    this._findIntervalIndices();
  }

  _findIntervalIndices() {
    const baseIntervalStrings = new Set(
      this._timelineCred.intervals().map(stringify)
    );
    const viewIntervalStrings = this._viewIntervals.map(stringify);
    for (const v of viewIntervalStrings) {
      if (!baseIntervalStrings.has(v)) {
        throw new Error(`invariant violation: can't find interval ${v}`);
      }
    }
    this._startIntervalIndex = Array.from(baseIntervalStrings).indexOf(
      viewIntervalStrings[0]
    );
    this._endIntervalIndex = Array.from(baseIntervalStrings).indexOf(
      viewIntervalStrings[viewIntervalStrings.length - 1]
    );

    if (this._startIntervalIndex < 0 || this._endIntervalIndex < 0) {
      throw new Error(`invariant violation: input interval 
        ${stringify(viewIntervalStrings[0]) +
          "-" +
          stringify(viewIntervalStrings[viewIntervalStrings.length - 1])} 
        is not contained in base intervals`);
    }
  }

  intervals(): $ReadOnlyArray<Interval> {
    return this._viewIntervals;
  }

  /**
   * Get the CredNode for a given NodeAddress.
   *
   * Returns undefined if the node is not in the filtered results.
   *
   * Note that it's possible that the node is present in the Graph, but not the
   * filtered results; if so, it will return undefined.
   */
  credNode(a: NodeAddressT): ?CredNode {
    const fullView = this._timelineCred.cred(a);
    if (fullView === undefined || fullView === null) {
      return undefined;
    }

    const cred = fullView.slice(
      this._startIntervalIndex,
      this._endIntervalIndex + 1
    );

    const total = sum(cred);
    const node = NullUtil.get(this._timelineCred.node(a));
    return {cred, total, node};
  }

  /**
   * Return all the nodes matching the prefix, along with their cred,
   * sorted by total cred (descending).
   */
  credSortedNodes(prefix: NodeAddressT): $ReadOnlyArray<CredNode> {
    const match = (a) => NodeAddress.hasPrefix(a, prefix);
    const addresses: NodeAddressT[] = this._timelineCred
      .availableNodes()
      .filter(match);
    const credNodes = addresses.map((a) => this.credNode(a));
    return sortBy(credNodes, (x: CredNode) => -x.total);
  }

  _getDefaultInterval(tc: TimelineCred): Interval {
    // weekIntervals is inclusive of start and end time
    return {
      startTimeMs: tc.intervals()[0].startTimeMs,
      endTimeMs: tc.intervals()[tc.intervals().length - 1].endTimeMs - 1,
    };
  }
}
