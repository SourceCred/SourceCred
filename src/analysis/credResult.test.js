// @flow

import stringify from "json-stable-stringify";
import {Graph, NodeAddress, EdgeAddress} from "../core/graph";
import * as Weights from "../core/weights";
import type {NodeType, EdgeType} from "./types";
import type {PluginDeclaration} from "./pluginDeclaration";
import {defaultParams} from "./timeline/params";
import {compute, toJSON, fromJSON} from "./credResult";

describe("analysis/credResult", () => {
  describe("to/fro JSON", () => {
    async function example() {
      const nodeType: NodeType = {
        name: "node",
        pluralName: "nodes",
        prefix: NodeAddress.fromParts(["node"]),
        defaultWeight: 2,
        description: "a type",
      };
      const edgeType: EdgeType = {
        forwardName: "points",
        backwardName: "is pointed to",
        prefix: EdgeAddress.fromParts(["edge"]),
        defaultWeight: {forwards: 2, backwards: 3},
        description: "a type",
      };
      const declaration: PluginDeclaration = {
        name: "empty",
        nodePrefix: NodeAddress.empty,
        edgePrefix: EdgeAddress.empty,
        nodeTypes: [nodeType],
        edgeTypes: [edgeType],
        userTypes: [nodeType],
      };

      const graph = new Graph()
        .addNode({
          address: NodeAddress.fromParts(["node", "1"]),
          description: "n1",
          timestampMs: 100000,
        })
        .addNode({
          address: NodeAddress.fromParts(["node", "2"]),
          description: "n2",
          timestampMs: 110000,
        });
      const weights = Weights.empty();
      const wg = {graph, weights};
      const params = defaultParams();
      const result = await compute(wg, params, [declaration]);
      return result;
    }

    describe("to/froJSON", () => {
      it("round-trips appropriately", async () => {
        const r = await example();
        const j = toJSON(r);
        const r_ = fromJSON(j);
        const j_ = toJSON(r_);

        expect(r.weightedGraph.graph.equals(r_.weightedGraph.graph)).toBe(true);
        expect(r.weightedGraph.weights).toEqual(r_.weightedGraph.weights);
        expect(r.credData).toEqual(r_.credData);
        expect(r.params).toEqual(r_.params);
        expect(r.plugins).toEqual(r_.plugins);
        expect(j).toEqual(j_);
      });
      it("snapshots as expected", async () => {
        const r = await example();
        const j = toJSON(r);
        expect(stringify(j, {space: 2})).toMatchSnapshot();
      });
    });
  });
});
