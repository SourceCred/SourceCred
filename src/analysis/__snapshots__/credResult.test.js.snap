// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analysis/credResult to/fro JSON to/froJSON snapshots as expected 1`] = `
"[
  {
    \\"type\\": \\"sourcecred/credResult\\",
    \\"version\\": \\"0.1.0\\"
  },
  {
    \\"credData\\": {
      \\"edgeOverTime\\": [
      ],
      \\"edgeSummaries\\": [
      ],
      \\"intervalEnds\\": [
        259200000
      ],
      \\"nodeOverTime\\": [
        {
          \\"cred\\": [
            0.5
          ],
          \\"seedFlow\\": [
            0.1
          ],
          \\"syntheticLoopFlow\\": [
            0.4
          ]
        },
        {
          \\"cred\\": [
            0.5
          ],
          \\"seedFlow\\": [
            0.1
          ],
          \\"syntheticLoopFlow\\": [
            0.4
          ]
        }
      ],
      \\"nodeSummaries\\": [
        {
          \\"cred\\": 0.5,
          \\"seedFlow\\": 0.1,
          \\"syntheticLoopFlow\\": 0.4
        },
        {
          \\"cred\\": 0.5,
          \\"seedFlow\\": 0.1,
          \\"syntheticLoopFlow\\": 0.4
        }
      ]
    },
    \\"params\\": {
      \\"alpha\\": 0.2,
      \\"intervalDecay\\": 0.5
    },
    \\"plugins\\": [
      {
        \\"type\\": \\"sourcecred/pluginDeclarations\\",
        \\"version\\": \\"0.1.0\\"
      },
      [
        {
          \\"edgePrefix\\": \\"E\\\\u0000\\",
          \\"edgeTypes\\": [
            {
              \\"backwardName\\": \\"is pointed to\\",
              \\"defaultWeight\\": {
                \\"backwards\\": 3,
                \\"forwards\\": 2
              },
              \\"description\\": \\"a type\\",
              \\"forwardName\\": \\"points\\",
              \\"prefix\\": \\"E\\\\u0000edge\\\\u0000\\"
            }
          ],
          \\"name\\": \\"empty\\",
          \\"nodePrefix\\": \\"N\\\\u0000\\",
          \\"nodeTypes\\": [
            {
              \\"defaultWeight\\": 2,
              \\"description\\": \\"a type\\",
              \\"name\\": \\"node\\",
              \\"pluralName\\": \\"nodes\\",
              \\"prefix\\": \\"N\\\\u0000node\\\\u0000\\"
            }
          ],
          \\"userTypes\\": [
            {
              \\"defaultWeight\\": 2,
              \\"description\\": \\"a type\\",
              \\"name\\": \\"node\\",
              \\"pluralName\\": \\"nodes\\",
              \\"prefix\\": \\"N\\\\u0000node\\\\u0000\\"
            }
          ]
        }
      ]
    ],
    \\"weightedGraph\\": [
      {
        \\"type\\": \\"sourcecred/weightedGraph\\",
        \\"version\\": \\"0.1.0\\"
      },
      {
        \\"graphJSON\\": [
          {
            \\"type\\": \\"sourcecred/graph\\",
            \\"version\\": \\"0.8.0\\"
          },
          {
            \\"edges\\": [
            ],
            \\"nodes\\": [
              {
                \\"description\\": \\"n1\\",
                \\"index\\": 0,
                \\"timestampMs\\": 100000
              },
              {
                \\"description\\": \\"n2\\",
                \\"index\\": 1,
                \\"timestampMs\\": 110000
              }
            ],
            \\"sortedNodeAddresses\\": [
              [
                \\"node\\",
                \\"1\\"
              ],
              [
                \\"node\\",
                \\"2\\"
              ]
            ]
          }
        ],
        \\"weightsJSON\\": [
          {
            \\"type\\": \\"sourcecred/weights\\",
            \\"version\\": \\"0.2.0\\"
          },
          {
            \\"edgeWeights\\": {
            },
            \\"nodeWeights\\": {
            }
          }
        ]
      }
    ]
  }
]"
`;
